include/kernel.h

```diff
@@ -800,14 +800,37 @@ void k_thread_system_pool_assign(struct k_thread *thread);
 /**
  * @brief Put the current thread to sleep.
  *
- * This routine puts the current thread to sleep for @a duration milliseconds.
+ * This routine puts the current thread to sleep for @a duration.
  *
- * @param ms Number of milliseconds to sleep.
+ * @param ms Duration of sleep, specified via a k_timeout_t
+ *           initialized via one of the K_TIMEOUT_*() macros.
+ *
+ * @return Zero if the requested time has elapsed or the number of
+ * system ticks left to sleep, if thread was woken up by \ref k_wakeup
+ * call.
+ */
+__syscall k_ticks_t k_sleep(k_timeout_t ms);
+
+/**
+ * @brief Sleep for ms timeout.
+ *
+ * This routine puts the current thread to sleep for @a duration
+ * milliseconds.  Simple wrapper around k_sleep().  Note that
+ * milliseconds may not match the rate of the internal tick counter
+ * and so this API can suffer from accidental precision loss where the
+ * k_timeout_t provides more control over units.
+ *
+ * @param ms Duration of sleep, specified in milliseconds
  *
  * @return Zero if the requested time has elapsed or the number of milliseconds
  * left to sleep, if thread was woken up by \ref k_wakeup call.
  */
-__syscall s32_t k_sleep(s32_t ms);
+static inline k_ticks_t k_msleep(s32_t ms)
```

Comment:
if function returns `number of milliseconds  left to sleep` then should that be same type as input parameter?

Reply:
The thinking was about compatibility: k_sleep() has always taken a 32 bit word and I didn't want to make that mutable via configuration in a way that could cause warnings in existing code (having been transformed to use k_msleep() of course).   But the return value is clearly going to be used (if it's used at all) as a timeout parameter, potentially to another API or in math with timeouts received via other APIs.   And that should have the same type as other timeouts.

I thought this was the lowest risk option, but I'd be OK with any combination of arguments really.  It could take a k_ticks_t or return a s32_t if someone has strong opinions.

Reply:
Hm... that's too many words.  Let me simplify:

It's an API that exists for compatibility.  So it should take the same argument that old code expects, because that's what it always got.  But it should return a value that new code expects so we don't get truncation errors as it evolves.

But I'm not a zealot.